File: FlutterUI/flutter_ui/lib/core/interceptors/authInterceptor.dart
import 'package:flutter_ui/core/models/response/singleResponseModel.dart';
import 'package:flutter_ui/core/models/user/refreshTokenModel.dart';
import 'package:flutter_ui/core/models/user/tokenModel.dart';
import 'package:flutter_ui/core/utilities/dependencyResolver.dart';
import 'package:flutter_ui/environments/api.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

class AuthInterceptor extends http.BaseClient {
  http.BaseClient _httpClient;

  AuthInterceptor() {
    _httpClient = http.Client();
  }

  @override
  Future<http.StreamedResponse> send(http.BaseRequest request) async {
    var tokenModel = await getTokens();
    var lang = await getLang();
    Map<String, String> headers;

    headers = {
      "content-type": "application/json",
      'Authorization': "Bearer " + "${tokenModel.token}",
      'lang': "$lang",
      "RefreshToken": "${tokenModel.refreshToken.refreshTokenValue}",
      "ClientId": "${tokenModel.refreshToken.clientId}",
      "ClientName": "${tokenModel.refreshToken.clientName}",
    };

    request.headers.addAll(headers);

    if (tokenModel.refreshToken.refreshTokenValue != null &&
        await tokenService.isTokenExpired()) {
      var newRequest = await _httpClient.post(
        Uri.parse(Environments.API_URL + "auth/refreshtoken"),
        headers: headers,
      );

      var response = SingleResponseModel<TokenModel>.fromJson(newRequest);
      if (response.success) {
        authService.onRefreshTokenSucceed(response.data);
        await tokenService.setToken(response.data.token);
        await tokenService
            .setRefreshToken(response.data.refreshToken.refreshTokenValue);
        await tokenService.setClientId(response.data.refreshToken.clientId);
      } else {
        authService.onRefreshTokenFailed();
        await tokenService.removeRefreshToken();
        await tokenService.removeToken();
      }
    }

    return await _httpClient.send(request);
  }

  Future<TokenModel> getTokens() async {
    var prefs = await SharedPreferences.getInstance();

    var token = await prefs.get("token");
    var refreshToken = await prefs.get("refresh-token");
    var clientId = await prefs.get("client-id");

    if (token != null) token = cryptoService.get(token);
    if (refreshToken != null) refreshToken = cryptoService.get(refreshToken);
    if (clientId != null) clientId = cryptoService.get(clientId);

    return TokenModel(
      token,
      null,
      RefreshTokenModel(
        refreshToken,
        null,
        clientId,
        Environments.CLIENT_NAME,
      ),
    );
  }

  Future<String> getLang() async {
    var prefs = await SharedPreferences.getInstance();

    var lang = await prefs.get("lang");
    if (lang != null) lang = cryptoService.get(lang);

    return lang;
  }
}

File: AngularUI/angular-ui/src/app/core/interceptors/auth.interceptor.ts
import { TokenService } from './../services/token.service';
import { Injectable } from '@angular/core';
import {
  HttpRequest,
  HttpHandler,
  HttpEvent,
  HttpInterceptor,
  HttpErrorResponse,
  HttpClient,
} from '@angular/common/http';
import { BehaviorSubject, Observable, throwError } from 'rxjs';
import { apiUrl, clientName } from 'src/api';
import { AuthService } from '../services/auth.service';
import { catchError, filter, switchMap, take, tap } from 'rxjs/operators';
import { Router } from '@angular/router';

@Injectable()
export class AuthInterceptor implements HttpInterceptor {
  token: string;
  refreshToken: string;
  clientId: string;

  constructor(
    private tokenService: TokenService,
    private authService: AuthService,
    private router: Router,
    private httpClient: HttpClient
  ) {}

  intercept(
    req: HttpRequest<any>,
    next: HttpHandler
  ): Observable<HttpEvent<any>> {
    this.token = this.tokenService.getToken();
    this.refreshToken = this.tokenService.getRefreshToken();
    this.clientId = this.tokenService.getClientId();

    req = this.addToken(req);

    return next.handle(req).pipe(
      catchError((error) => {
        if (error instanceof HttpErrorResponse && error.status === 401) {
          this.router.navigate(['/auth/login']);
          return throwError(error);
        } else if (error instanceof HttpErrorResponse && error.status === 403) {
          if (this.refreshToken != null) {
            return this.handle403Error(req, next);
          }

          return throwError(error);
        } else {
          return throwError(error);
        }
      })
    );
  }
  private addToken(req: HttpRequest<any>) {
    return req.clone({
      setHeaders: {
        Authorization: `Bearer ${this.token ?? ''}`,
        lang: localStorage.getItem('lang') ?? '',
        ClientId: this.clientId ?? '',
        RefreshToken: this.refreshToken ?? '',
        ClientName: clientName,
      },
    });
  }

  private isRefreshing = false;
  private refreshTokenSubject: BehaviorSubject<any> = new BehaviorSubject<any>(
    null
  );

  private handle403Error(
    request: HttpRequest<any>,
    next: HttpHandler
  ): Observable<HttpEvent<any>> {
    if (!this.isRefreshing) {
      this.isRefreshing = true;
      this.refreshTokenSubject.next(null);

      return this.httpClient
        .post<any>(
          apiUrl + 'auth/refreshToken',
          {},
          {
            headers: {
              Authorization: 'Bearer ' + this.token ?? '',
              lang: localStorage.getItem('lang') ?? '',
              clientId: this.clientId ?? '',
              refreshToken: this.refreshToken ?? '',
              clientName: clientName,
            },
          }
        )
        .pipe(
          tap((response) => {
            if (response.success) {
              this.tokenService.setToken(response.data.token);
              this.tokenService.setRefreshToken(
                response.data.refreshToken.refreshTokenValue
              );
              this.tokenService.setClientId(
                response.data.refreshToken.clientId
              );
              this.authService.onRefreshTokenSucceed(response.data);
            }
          }),
          catchError((error) => {
            this.tokenService.removeRefreshToken();
            this.tokenService.removeToken();
            this.authService.onRefreshTokenFailed();
            return throwError(error);
          })
        )
        .pipe(
          switchMap((response: any) => {
            this.tokenService.setToken(response.data.token);
            this.tokenService.setRefreshToken(
              response.data.refreshToken.refreshTokenValue
            );
            this.tokenService.setClientId(response.data.refreshToken.clientId);
            this.authService.onRefreshTokenSucceed(response.data);

            this.isRefreshing = false;
            this.refreshTokenSubject.next(response.data.token);
            return next.handle(this.addToken(request));
          }),
          catchError((error) => {
            this.tokenService.removeRefreshToken();
            this.tokenService.removeToken();
            this.authService.onRefreshTokenFailed();
            return throwError(error);
          })
        );
    } else {
      return this.refreshTokenSubject.pipe(
        filter((token) => token != null),
        take(1),
        switchMap((jwt) => {
          return next.handle(this.addToken(request));
        })
      );
    }
  }
}
